### PR TITLE
[diffusion, reward, trainer, recipe] feat: add LTX-2.3 OmniNFT training

### LINKED ISSUE #414: [RFC] Support OmniNFT Modality-Decoupled Advantage Routing for LTX-2.3
## Feature request

Add support for the first-stage OmniNFT algorithm (https://arxiv.org/abs/2605.12480) in `verl-omni`, using LTX-2.3 as the target joint audio-video generation model.

The initial implementation focuses on OmniNFT's modality-decoupled advantage routing and reuses the existing LTX-2.3 rollout and DiffusionNFT training infrastructure.

Status: proposal.

## Motivation

Joint audio-video generation produces multiple rewards with different scopes:

- visual quality rewards evaluate the video branch;
- audio quality rewards evaluate the audio branch;
- audio-video synchronization rewards evaluate both branches.

Using one globally aggregated advantage can assign unrelated reward signals to the wrong modality. OmniNFT computes an independent advantage for each reward and routes it to the corresponding video/audio loss.

This contribution would provide a practical online multi-reward DiffusionNFT recipe for LTX-2.3 and establish a reusable reward-routing abstraction for similar multimodal diffusion models.

### Goals

- Support online DiffusionNFT training for LTX-2.3.
- Reuse the existing `vllm-omni` LTX-2.3 joint audio-video rollout.
- Export the clean video/audio latents and replay conditions required by the actor.
- Support five native reward functions with batched local inference:
  - VideoAlign;
  - HPSv3;
  - Audiobox Aesthetics;
  - CLAP;
  - DeSync/Synchformer.
- Support colocated reward models with load/offload scheduling.
- Compute per-reward group advantages without prematurely aggregating reward scores.
- Route advantages independently to the video and audio NFT losses.
- Support FSDP2 and context-parallel training.
- Validate the implementation on `train_metadata_20k.jsonl` and JavisBench-mini.

### Non-goals

- Layer-wise gradient surgery.
- Region-wise loss reweighting.
- Model-specific modification of LTX attention layers.

The excluded OmniNFT components depend on model-specific cross-attention structure, layer semantics, and token mappings. They are therefore reserved for a later model-specific extension after the basic RL loop has been validated.

## Design

The proposed training flow is:

```text
train_metadata_20k.jsonl
  -> prompt groups
  -> vLLM-Omni LTX-2.3 rollout
  -> decoded audio/video samples
  -> five native batched rewards
  -> per-reward group advantages
  -> modality-specific advantage routing
  -> video/audio DiffusionNFT losses
  -> FSDP2 + CP actor update
  -> synchronize default and old policies
```

For each reward \(k\), compute an independent group advantage:

$$A_{i,k} = \frac{R_{i,k}-\mu_{g,k}}{\sigma_{g,k}+\epsilon}$$

The routed advantages are:

$$\widetilde A_{i,v} = A_{i,\mathrm{VideoAlign}} + A_{i,\mathrm{HPSv3}} + A_{i,\mathrm{DeSync}}$$
$$\widetilde A_{i,a} = A_{i,\mathrm{AudioBox}} + A_{i,\mathrm{CLAP}} + A_{i,\mathrm{DeSync}}$$

The resulting video and audio optimality probabilities are used independently by the two modality-specific DiffusionNFT losses.

The implementation should preserve the existing framework boundaries:

- `vllm-omni`: LTX-2.3 rollout and clean-latent export;
- `verl-omni`: data flow, reward scheduling, advantage routing, DiffusionNFT actor, policy lifecycle, FSDP2, and CP;
- native reward manager: colocated reward model initialization, batching, device placement, and load/offload.

## Reusable interfaces

The implementation should expose generic multimodal contracts rather than hard-code the five rewards into the trainer:

- named reward outputs with stable reward dimensions;
- reward-to-modality routing weights;
- separate video/audio optimality probabilities;
- modality-specific prediction and loss outputs;
- replayable multimodal latent and conditioning records.

This allows similar joint audio-video models to reuse the routing logic with a different reward-routing matrix.

## Validation

The contribution should include:

1. Contract tests for reward tensor shapes, independent normalization, and routing.
2. A single-batch test verifying that video rewards affect only the video loss, audio rewards affect only the audio loss, and synchronization rewards affect both.
3. LTX-2.3 rollout/replay consistency checks for clean latents and conditioning metadata.
4. A one-step FSDP2 + CP actor update with finite losses and gradients.
5. Native batched reward inference with all five reward models.
6. A short training run on `train_metadata_20k.jsonl`.
7. Evaluation of the base and trained checkpoints on the same JavisBench-mini prompts, seeds, resolution, frame count, and evaluator version.

The primary result should compare the routed implementation against a vanilla aggregated-reward DiffusionNFT baseline using the same rollout and evaluation protocol.

## Dependencies and risks

- LTX-2.3 rollout must expose sufficient clean-latent and conditioning metadata for actor replay.
- The five native reward models must fit the colocated device schedule through batching and load/offload.
- Video and audio prediction tensors must remain distinguishable through the training loss.
- FSDP2 and CP must preserve modality-specific masks and reductions.
- Reward normalization and modality weights may require calibration to avoid one modality dominating the update.
- The first version intentionally evaluates the one-stage LTX-2.3 pipeline and should report relative improvement over its own baseline rather than direct reproduction of OmniNFT paper scores.

## Implementation plan

- Add the LTX-2.3 OmniNFT rollout/replay adapter.
- Add multimodal reward records and batched native reward execution.
- Add per-reward advantage computation and configurable modality routing.
- Extend the DiffusionNFT actor to compute separate video/audio losses.
- Add FSDP2 and CP validation.
- Add the LTX-2.3 OmniNFT recipe, dataset configuration, and evaluation commands.
- Document known limitations and the path for future gradient-surgery and region-reweighting extensions.

## Contribution

I can provide the LTX-2.3 rollout/training adapters, native multimodal reward recipe, routing implementation, contract tests, and short-run validation results.

### PR DESCRIPTION
**Pull request by DuwSpace**
### What does this PR do?

Adds online OmniNFT training for LTX-2.3 audio-video generation:

- five native reward adapters: VideoAlign, HPSv3, AudioBox Aesthetics, CLAP, and DeSync;
- modality-wise advantage routing into independent video and audio probabilities;
- joint LTX-2.3 old/current/reference forwards and paired video/audio DiffusionNFT losses;
- a reproducible Ascend NPU recipe with pinned data/model preparation, GitHub-hosted reference/trained video comparisons, result curves, CPU coverage, and a tiny end-to-end smoke test.

The implementation is an online direct-preference path: rollout returns final clean video/audio latents and replay conditions, rewards remain separate through group normalization, and the actor performs one joint backward pass over the routed video/audio objectives.

### Paper scope

This PR implements one of the three techniques described by OmniNFT:

| Technique | Status |
| --- | --- |
| Modality-wise advantage routing | Implemented |
| Layer-wise gradient surgery | Not implemented |
| Region-wise loss reweighting | Not implemented |

The other two paper techniques are outside the scope of RFC #414 and this PR.

### Related issue

Closes #414.

### Duplicate-work check

- [x] Read #414 and its ownership/design discussion on September 18, 2026.
- [x] Searched open PRs for `414 in:body`.
- [x] Searched open PRs for `OmniNFT LTX-2.3 in:title,body`.
- [x] No duplicate open PR was found.

#### NPU evidence

- The batch-32 experiment ran for 196 training steps with validation every 10 steps. Step 196 recorded actor total loss `3.328362` and gradient norm `0.015320`. Each training reward curve shows its raw values and 30-step trailing mean; validation rewards are shown as raw points joined by dashed lines.

**Training rewards**

Batch-32 training rewards

**Validation rewards**

Batch-32 validation rewards

#### Woman

<details>
<summary>Prompt</summary>

In a medium close-up, a young woman with blonde shoulder-length hair stands in a lavender field under a twilight sky. She says, "告诉我这条光滑的绿色带子见证了多少年的沉重。" The audio features gentle, atmospheric singing establishing a calm and wistful mood.

</details>

| Reference | Trained |
|:---:|:---:|
| <video src="https://github.com/user-attachments/assets/a8b4ddff-41ae-4221-b985-23e910647244" controls></video> | <video src="https://github.com/user-attachments/assets/17c2e67a-a97f-48a5-a1c4-0396518d1c07" controls></video> |

#### Piano

<details>
<summary>Prompt</summary>

In a studio with beige brick walls and a light gray floor, a man wearing a black fedora, a white T-shirt, and dark jeans sits on a brown leather bench. He leans forward, his hands positioned to play a wooden upright piano in a medium shot captured from a slight high angle. Two condenser microphones on boom arms are positioned above the piano's open lid, and professional lighting equipment, including a black floor lamp and several stands, is arranged around him, casting a warm glow on the scene. From this perspective, the man's hands move with focus across the piano keys. Below the keyboard, his feet in black sneakers actively work the instrument's pedals as well as a pedalboard with effects units resting on the floor. His performance unfolds within the initial framing, with the microphones, brick wall texture, and surrounding equipment remaining visible throughout the shot. The audio shows soft piano notes being played, filling the space with the sound of the performance.

</details>

| Reference | Trained |
|:---:|:---:|
| <video src="https://github.com/user-attachments/assets/e4ebebe2-e060-45fa-9024-3818bb7349f9" controls></video> | <video src="https://github.com/user-attachments/assets/71319c70-e66d-4edb-84e1-d4b1561e90c8" controls></video> |

#### Humanoid

<details>
<summary>Prompt</summary>

An extreme close-up centers on a whimsical humanoid figure made entirely of green leafy vegetables under bright, natural daylight. His skin and hair are textured like kale, while broccoli forms his neck and shoulders. With large white eyes and an open mouth, he holds both arms raised outward in a garden setting, where blurred green foliage creates a soft, out-of-focus background. The figure's leafy hands gesture animatedly toward the viewer. His arched eyebrows lift higher, and his eyes widen in an expression of surprise. He then leans forward, bringing his face closer to the camera, his expression shifting to one of earnest eagerness as he begins to speak. He says, "Oh, how did you find me? Although it is a bit inappropriate, for the sake of your health, I hope you remember to eat more vegetables." The audio shows only the character's spoken dialogue. No other ambient sound, music, or sound effects are present.

</details>

| Reference | Trained |
|:---:|:---:|
| <video src="https://github.com/user-attachments/assets/ad4c7d46-0204-4062-80ae-b52f0ea0b804" controls></video> | <video src="https://github.com/user-attachments/assets/9ca3a3c2-f09a-4a29-924d-cea7c4347ad4" controls></video> |

#### Anime

<details>
<summary>Prompt</summary>

In a vibrant, anime-style 3D game world, a female character with blonde hair, wearing a white and blue coat with red boots, walks away from the camera across a vast expanse of violet grassland. The scene is lit by the warm oranges and yellows of a sunset sky, where a large, rocky mountain in the distance erupts with fire and smoke. A long, green leaf-like object extends horizontally across the middle of the frame, above a calm body of water and fields of purple and blue flora. A user interface displays “一只飞来洲 bilibili” beside a cartoon dragon logo in the top-left corner and “尊贵的DJ机主” next to a character portrait and a boxed number “1” in the top-right. In the bottom-right, a green circular button holds a white mushroom symbol. The character's walking pace quickens to a run as swirling green energy, representing a wind-element skill, gathers around her hands in visible gusts and particles. As she advances, several fantastical creatures suddenly emerge from the ground ahead of her, their forms materializing amidst the purple grass. With a swift forward gesture, she unleashes a powerful whirlwind that erupts from her position. The vortex of air and debris immediately lifts the creatures, tumbling them violently in the air. As they are caught in the whirlwind, red health bars appear above each monster and deplete rapidly. The surrounding violet grass and uniquely shaped trees bend and sway violently in the powerful gusts emanating from her attack, while the mountain in the background continues to spew fire and smoke. The audio shows a deep, continuous rumble of distant explosions echoing from the fiery mountains. This is layered with the sounds of the character's footsteps on the grass, a charging whoosh as the wind skill activates, and the powerful, gusting roar of the ensuing whirlwind. The violent swaying of flora adds a rustling texture to the soundscape, alongside UI sound effects indicating the rapid depletion of the creatures' health bars.

</details>

| Reference | Trained |
|:---:|:---:|
| <video src="https://github.com/user-attachments/assets/a8ac6edc-6720-4c49-b003-7bfa954208da" controls></video> | <video src="https://github.com/user-attachments/assets/eaec5a13-7b90-4301-9e38-a99f2ba83933" controls></video> |

#### Soldiers

<details>
<summary>Prompt</summary>

In a wide shot with a desaturated, cinematic quality, six soldiers in dark World War II-era British military uniforms and helmets walk away from the camera down the center of a narrow, paved street. The scene is captured with deep focus under diffuse, overcast daylight. Leaflets and small papers drift steadily from the sky, scattering across the road amidst other debris. The street is flanked by two-story brick and plaster houses with chimneys and utility lines, and a seventh soldier stands on the left sidewalk, facing slightly toward the group. The soldiers, with rifles slung over their shoulders and some carrying backpacks, wander slowly forward through the eerily empty setting. Suddenly, their aimless pace changes. They slow their walk and their bodies tense, their posture shifting into a slight crouch. Gripping their rifles tightly, they bring the weapons from their shoulders into a ready position, aiming them forward down the street and heightening the tension of the moment. The audio shows a quiet atmosphere dominated by the faint sound of wind and distant, intermittent bird calls. This stillness is abruptly broken by a series of explosions that echo from afar, their reverberations cutting through the silence.

</details>

| Reference | Trained |
|:---:|:---:|
| <video src="https://github.com/user-attachments/assets/00e2f395-6d2c-477a-9268-1f791a33f1a0" controls></video> | <video src="https://github.com/user-attachments/assets/c7ac2996-5ed5-421c-879d-0e6c0f9e21f6" controls></video> |

#### Sailing

<details>
<summary>Prompt</summary>

17th century sailing ship making a path through the waves during a storm.

</details>

| Reference | Trained |
|:---:|:---:|
| <video src="https://github.com/user-attachments/assets/e3cda433-fc61-4d6d-8861-e060ec5f6676" controls></video> | <video src="https://github.com/user-attachments/assets/9cb16ab8-3b62-4901-9969-1fc1ec5b9a76" controls></video> |

### API and usage example

```bash
python3 examples/omninft_trainer/data_process/prepare_data.py
bash examples/omninft_trainer/download_models.sh
WANDB_MODE=offline \
  bash examples/omninft_trainer/ltx2/run_ltx2_3_omninft_lora_npu_bs32.sh
```

The checked-in recipe uses batch size 32, learning rate `3e-5`, LoRA rank/alpha 32/64, four TP=4 rollout replicas on 16 NPUs, and five explicitly placed native reward groups. Paths and placement remain environment-overridable.

Co-author: @shuy98 (Shu You).

AI assistance from OpenAI Codex was used for implementation, testing, documentation, and commit-history reconstruction.

### CODE REVIEW

**Comment by NancyFyong** [comment]
thx for your contribution, I think we can spilit it to several small pr to review?

File: examples/omninft_trainer/assets/omninft_bs32_step196_loss.svg
Comment (NancyFyong) [inline examples/omninft_trainer/assets/omninft_bs32_step196_loss.svg]: we can put it in readme

File: verl_omni/utils/reward_score/clap_native.py
Comment (ZihaoW123) [inline verl_omni/utils/reward_score/clap_native.py:1]: “What does "_native" mean? Can you reuse the existing "clap" and "hpsv3" in the repository?”

**Comment by shuy98** [comment]
> thx for your contribution, I think we can spilit it to several small pr to review?

sure we can do that

File: verl_omni/utils/reward_score/clap_native.py
Comment (shuy98) [reply verl_omni/utils/reward_score/clap_native.py:1]: > “What does "_native" mean? Can you reuse the existing "clap" and "hpsv3" in the repository?”

that's right, we will do a refactoring on that

**Comment by shuy98** [comment]
@NancyFyong @ZihaoW123  the splitted PR has been attached #663 #664 , and native reward has been refactored, please check, thanks!
